src/partition/repart.c

```diff
@@ -1811,40 +1813,42 @@ static int context_load_partition_table(
         }
 
 add_initial_free_area:
+        sector_size = fdisk_get_sector_size(c);
+
         nsectors = fdisk_get_nsectors(c);
-        assert(nsectors <= UINT64_MAX/512);
-        nsectors *= 512;
+        assert(nsectors <= UINT64_MAX/sector_size);
+        nsectors *= sector_size;
 
         first_lba = fdisk_get_first_lba(c);
-        assert(first_lba <= UINT64_MAX/512);
-        first_lba *= 512;
+        if (sector_size == 512)
+                first_lba = round_up_size(first_lba, 8);
+        assert(first_lba <= UINT64_MAX/sector_size);
+        first_lba *= sector_size;
 
         last_lba = fdisk_get_last_lba(c);
         assert(last_lba < UINT64_MAX);
         last_lba++;
-        assert(last_lba <= UINT64_MAX/512);
-        last_lba *= 512;
+        if (sector_size == 512)
+                last_lba = round_down_size(last_lba, 8);
+        assert(last_lba <= UINT64_MAX/sector_size);
+        last_lba *= sector_size;
 
         assert(last_lba >= first_lba);
 
         if (left_boundary == UINT64_MAX) {
                 /* No partitions at all? Then the whole disk is up for grabs. */
 
-                first_lba = round_up_size(first_lba, 4096);
-                last_lba = round_down_size(last_lba, 4096);
-
                 if (last_lba > first_lba) {
                         r = context_add_free_area(context, last_lba - first_lba, NULL);
                         if (r < 0)
                                 return r;
                 }
         } else {
                 /* Add space left of first partition */
-                assert(left_boundary >= first_lba);
+                if (sector_size == 512)
+                        left_boundary = round_down_size(left_boundary, 4096);
 
-                first_lba = round_up_size(first_lba, 4096);
-                left_boundary = round_down_size(left_boundary, 4096);
-                last_lba = round_down_size(last_lba, 4096);
+                assert(left_boundary >= first_lba);
```

Comment:
It seems right to also perform this assertion check after aligning, but I'm not really sure if it is.